### PR TITLE
Object.getHash() shouldn't be casting away const

### PR DESCRIPTION
**Pull request by WalterBright**
blocking https://github.com/dlang/dmd/pull/14164

### CODE REVIEW

File: src/core/thread/fiber.d [hunk 1]
```diff
@@ -605,12 +605,12 @@ class Fiber
           size_t guardPageSize = PAGESIZE ) nothrow
     in
     {
-        assert( dg );
+        assert( cast(void delegate() const) dg );
```
Comment (dkorpel) [inline src/core/thread/fiber.d:608]: Why is this cast needed? It looks like a simple null check.

Comment (WalterBright) [reply src/core/thread/fiber.d:608]: It isn't a simple null check. It winds up calling dassert() which has overloads for delegates.

**Review by dkorpel — APPROVED** [review]

File: src/core/thread/fiber.d (hunk 1 above)
Comment (dkorpel) [reply src/core/thread/fiber.d:608]: That looks like an issue on its own, `assert(dg)` should just compile. Is there a bugzilla issue for that or am I missing something?

Comment (WalterBright) [reply src/core/thread/fiber.d:608]: assert's inside contracts behave differently.

**Comment by WalterBright** [comment]
The buildkite errors don't seem to be related to this PR.

**Comment by WalterBright** [comment]
It's hard to see how the buildkite errors have anything to do with this PR.

File: src/core/thread/fiber.d (hunk 1 above)
Comment (ljmf00) [reply src/core/thread/fiber.d:608]: > assert's inside contracts behave differently.

Can you elaborate?

Comment (WalterBright) [reply src/core/thread/fiber.d]: Here's the error if I leave it unchanged:
```
src/core/thread/fiber.d(609): Error: none of the overloads of template `core.internal.dassert._d_assert_fail` are callable using argument types `!(void delegate())(string, void delegate())`
src/core/internal/dassert.d(38):        Candidates are: `_d_assert_fail(A)(scope const string op, auto ref scope const A a)`
src/core/internal/dassert.d(63):                        `_d_assert_fail(A...)`
```

Comment (ljmf00) [reply src/core/thread/fiber.d]: It looks like we should add that overload on druntime. Does it make sense to the user to cast it exolicitly to const?

**Comment by ljmf00** [comment]
> The buildkite errors don't seem to be related to this PR.

Yeah, it's not. See https://github.com/vibe-d/vibe.d/pull/2661 for context.

File: src/core/thread/fiber.d [hunk 2]
```diff
@@ -605,12 +605,13 @@ class Fiber
           size_t guardPageSize = PAGESIZE ) nothrow
     in
     {
-        assert( dg );
+//        assert( cast(void delegate() const) dg );
+//        assert( dg );
```
Comment (RazvanN7) [inline src/core/thread/fiber.d:609]: Why not just delete this code instead of leaving it commented?

Comment (WalterBright) [reply src/core/thread/fiber.d:609]: Because I don't know what is triggering the buildkite failures, or if those failures are unrelated to this PR.

Comment (dkorpel) [reply src/core/thread/fiber.d:609]: The vibe-d and mir-algorithm failures are unrelated and being worked on

**Comment by WalterBright** [comment]
buildkike/vibe is still failing, although the libmir problem has disappeared.

**Comment by Geod24** [comment]
Vibe.d has been fixed. Please remove the code instead of commenting it tho.

**Comment by WalterBright** [comment]
It's removed now.

**Pull request merged**